rust/kernel/thread.rs

```diff
@@ -0,0 +1,198 @@
+    /// let mut a = 1;
+    ///
+    /// let t = Thread::new(
+    ///   move || {
+    ///     let b = Box::try_new(42)?;
+    ///
+    ///     for _ in 0..10 {
+    ///         a = a + 1;
+    ///         println!("Hello Rust Thread {}", a + b.as_ref());
+    ///     }
+    ///     Ok(())
+    ///   },
+    ///   cstr!("rust-thread")
+    /// )?;
+    ///
+    /// t.wake_up();
+    pub fn new<F>(f: F, name: CStr) -> KernelResult<Self>
+    where
+        F: FnOnce() -> KernelResult<()>,
+        F: Send + 'static,
+    {
+        // Allocate closure here, because this function maybe returns before `rust_thread_func`
+        // (the function that use the closure) get executed.
+        let bf: Box<dyn FnOnce() -> KernelResult<()> + 'static> = Box::try_new(f)?;
+
+        // Double boxing here because `dyn FnOnce` is a fat pointer, and we can only pass a usize
+        // as the `data` for `kthread_create_on_node`.
+        //
+        // We `into_raw` from this side, and will `from_raw` at the other side to transfer the
+        // ownership of the boxed data.
+        let db = Box::into_raw(Box::try_new(bf)?) as *mut _;
+
+        let task;
+
+        // SAFETY:
+        //
+        // * `db` is a proper pointer (generated by `Box::into_raw()`), and if succeed, the new
+        // thread will get the ownership.
+        //
+        // * `kthread_create_on_node` will copy the content of name, so we don't need to make the
+        // name live longer.
+        unsafe {
+            task = bindings::kthread_create_on_node(
+                Some(rust_thread_func),
+                db,
+                bindings::NUMA_NO_NODE,
+                "%s".as_ptr() as _,
+                name.as_ptr(),
+            );
+        }
+
+        let result = ptr_to_result(task);
+
+        if result.is_err() {
+            // Creation fails, we need to get back the double boxed closure.
+            //
+            // SAFETY:
+            //
+            // `db` is a proper pointer generated by a `Box::into_raw()` from a box created by us,
+            // if the thread creation fails, no one will consume that pointer.
+            unsafe {
+                Box::from_raw(db);
+            }
+        } else {
+            // Increase the refcount of the task, so that it won't go away if it `do_exit`.
+            //
+            // SAFETY:
+            //
+            // `task` is a valid pointer to newly created thread
+            unsafe {
+                rust_helper_get_task_struct(task);
+            }
+        }
+
+        Ok(Thread { task: result? })
+    }
+
+    /// Wake up the kernel thread, note that a newly created kthread will not run until `wake_up`
+    /// is called.
+    ///
+    /// This function might sleep, don't call in atomic contexts.
+    pub fn wake_up(&self) {
+        // SAFETY:
+        //
+        // `task` is a valid pointer to a kernel thread structure, and it's not stopped because
+        // `stop` will consume the [`Thread`].
```

Comment:
```suggestion
        // SAFETY: `task` is a valid pointer to a kernel thread structure, and it's not stopped because
        // `stop` will consume the [`Thread`].
```

Also, please try to keep comments under the 80 limit (normal code goes up to 100).

Reply:
Seems I missed to fix this, it's late now, I will send an update with this fixed.

Reply:
Which reminders me, I saw we increased the line width to 100 (80 is still prefered by scripts/checkpatch.pl will use 100) from:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=bdc48fa11e46f867ea4d75fa59ee87a7f48be144

So do we use the new limitation?

Reply:
We picked the `rustfmt` defaults. I initially thought the same, but on the other hand it is true that text can be easier to read when it has a limited width. I don't mind one way or the other. Same thing with indentation vs. the C code.

On the topic of 80 vs 100, it is never clear. I tried to raise the limit to 100 in `.clang-format`, and I was met with push back. Actually, we just had another patch for that too: https://lore.kernel.org/lkml/20210319183714.8463-1-ansuelsmth@gmail.com/

Reply:
For the `rustfmt` defaults, do we want to enable [wrap_comments](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#wrap_comments) option (which is disabled by default)? Otherwise `rustfmt` can not help us for comment line width.

Reply:
> For the `rustfmt` defaults, do we want to enable [wrap_comments](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#wrap_comments) option (which is disabled by default)? Otherwise `rustfmt` can not help us for comment line width.

Hmm.. except that `rustfmt` will wrap inside an intra link, for example, "\[\`Thread::try_new\`\]" can be wrapped right after the second colon symbol, and give us:
````
/// ................................................... [`Thread::
/// try_new`]
````
, which breaks the intra link doc generation. ;-(

I will do a manually `rustfmt --config wrap_comments=true` and try to fit all comments in to 80 columns.

Reply:
> I will do a manually `rustfmt --config wrap_comments=true` and try to fit all comments in to 80 columns.

Turns out that `rustfmt` won't wrap comments in the code (the comments begin with "//" not "///") to 80 columns even with `wrap_comments=true`. Should we then only limit documentation comments to 80 columns then?

Reply:
Yeah, sorry, I have to still finish the style guide for comments, do a pass on all the code, enforce it with the tidy script, etc. If you can fit this PR in the 80 limit, that would be great!

Reply:
Sure ;-)